src/Compilers/Core/Portable/SourceGeneration/Nodes/InputNode.cs

```diff
@@ -19,17 +19,19 @@ internal sealed class InputNode<T> : IIncrementalGeneratorNode<T>
     {
         private readonly Func<DriverStateTable.Builder, ImmutableArray<T>> _getInput;
         private readonly Action<IIncrementalGeneratorOutputNode> _registerOutput;
+        private readonly IEqualityComparer<T> _inputComparer;
```

Comment:
>_inputComparer [](http://example.com/codeflow?start=46&length=14)

When is `_inputComparer` needed? Isn't `CommonCompiler.ResolveAdditionalFilesFromArguments()` responsible for dropping duplicates?

Reply:
`ResolveAdditionalFilesFromArguments` drops duplicates that are passed in for a single compilation request. The `_inputComparer` compares between the de-duplicated inputs from this run compared with those from the previous one. 

In general we used reference equality to compare additional texts in the compiler, as they are only created once at the start of the compilation request. That's no longer true, because we create a file per-request and have to actually compare content to determine if the file has changed between invocations.